### ISSUE TITLE
Optional custom error for constraints?

### ISSUE THREAD

**Issue by fanatid** [project-serum/anchor#509]
Sometimes it's can be useful to have a special error code for constraints because this can provide a clear error message. Possible syntax:
```
#[account(constraint = <expression> ? MyError)]
```

Maybe not limited by constraint only?

**Comment by armaniferrante** [comment]
Agreed. I think this would be useful for any constraint, particularly those where there can be more than one, e.g., `has_one`, so that one can differentiate between the scenarios.

I like the `?`. Another possibility would be to use a `;`.

```
#[account(constraint = <expression>; MyError)]
```

Though I don't have a preference. Tag @Henry-E who I previously discussed this with as well.

**Comment by fanatid** [comment]
In the case of `;` how few constraints will look like?
```
#[account(
  constraint = <expression>; MyError,
  constraint = <expression>
)]
```
vs

```
#[account(
  constraint = <expression> ? MyError,
  constraint = <expression>
)]
```

but in general, I do not care about symbols if a custom error will be possible.

**Comment by armaniferrante** [comment]
Another one that I like in addition to `?` is

```
constraint = <expression> ! MyError
```

**Comment by Henry-E** [comment]
Not sure but there might be better error logging in general coming soon
which solves this problem without needing extra syntax

**Comment by rav3nh0lm** [comment]
Yes would be amazing, I am trying to debug some code, and it's incredibly exhausting trying to understand where the error originated from (i.e. from one of the macros)
